File: Sources/RedlineCore/RedlineHome.swift
// Where everything RedLine reads and writes hangs off.
//
// One helper rather than thirty calls to homeDirectoryForCurrentUser, for two reasons. The
// end to end tests need to drive the real binary against a directory that is not the real
// home, and macOS resolves the home directory from the account rather than from $HOME, so
// exporting HOME does not do it. And a second profile on one machine becomes possible
// without a second account.
//
// REDLINE_HOME is honoured only when it names an absolute path that already exists, because
// a typo that silently creates a fresh empty history somewhere is worse than being ignored.
import Foundation

public enum RedlineHome {
    public static let variable = "REDLINE_HOME"

    public static var url: URL {
        if let raw = ProcessInfo.processInfo.environment[variable],
           raw.hasPrefix("/"),
           FileManager.default.fileExists(atPath: raw) {
            return URL(fileURLWithPath: raw, isDirectory: true)
        }
        return FileManager.default.homeDirectoryForCurrentUser
    }

    /// True when the paths in use are not the account's own. Surfaced so a diagnostic can
    /// say so rather than leaving someone puzzled about missing history.
    public static var isOverridden: Bool {
        url.standardizedFileURL != FileManager.default.homeDirectoryForCurrentUser
            .standardizedFileURL
    }

    public static func path(_ components: String) -> URL {
        url.appendingPathComponent(components)
    }
}

File: windows/src/Redline.Core/RedlineHome.cs
// Where everything RedLine reads and writes hangs off. REDLINE_HOME moves it, but only when it
// names an absolute path that already exists, so a typo is ignored rather than starting a profile.
namespace Redline.Core;

public static class RedlineHome
{
    public const string Variable = "REDLINE_HOME";

    public static string AccountHome =>
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);

    public static string Url
    {
        get
        {
            var raw = Environment.GetEnvironmentVariable(Variable);
            if (!string.IsNullOrEmpty(raw) && Path.IsPathFullyQualified(raw) && Directory.Exists(raw))
                return Path.GetFullPath(raw);
            return AccountHome;
        }
    }

    /// True when the paths in use are not the account's own, so a diagnostic can say so.
    public static bool IsOverridden =>
        !string.Equals(Path.GetFullPath(Url).TrimEnd('\\', '/'),
                       Path.GetFullPath(AccountHome).TrimEnd('\\', '/'),
                       StringComparison.OrdinalIgnoreCase);

    /// Joins forward-slash relative components onto the home, e.g. ".local/share/redline".
    public static string PathFor(string components) => Join(Url, components);

    public static string Join(string root, string components) =>
        Path.Combine(new[] { root }.Concat(components.Split('/', StringSplitOptions.RemoveEmptyEntries)).ToArray());

    /// RedLine's own data directory: snapshot, feed, shim log, history.
    public static string DataDir(string? home = null) => Join(home ?? Url, ".local/share/redline");
}
